leafmap/ai.py

```
@@ -0,0 +1,774 @@
+        result = run_safe_code(code, output_path=output)
+        # 记录本次错误，供后续修复参考
+        history.append(
+            {
+                "attempt": attempt,
+                "error": result.get("error", ""),
+                "traceback": result.get("traceback", ""),
+            }
+        )
+
+    if result["ok"]:
+        # 成功案例缓存：保存本次成功的代码，下次相同描述直接复用
```

Ruff (0.16.1)

[warning] Comment contains ambiguous `：` (FULLWIDTH COLON). Did you mean `:` (COLON)?

(RUF003)